op-bindings/etherscan/client.go

```diff
@@ -0,0 +1,215 @@
+package etherscan
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-service/retry"
+)
+
+type client struct {
+	baseUrl string
+}
+
+type apiResponse struct {
+	Status  string          `json:"status"`
+	Message string          `json:"message"`
+	Result  json.RawMessage `json:"result"`
+}
+
+type rpcResponse struct {
+	JsonRpc string          `json:"jsonrpc"`
+	Id      int             `json:"id"`
+	Result  json.RawMessage `json:"result"`
+}
+
+type TxInfo struct {
+	TxHash string `json:"txHash"`
+	To     string `json:"to"`
+	Input  string `json:"input"`
+}
+
+const apiMaxRetries = 3
+const apiRetryDelay = time.Duration(2) * time.Second
+const errRateLimited = "Max rate limit reached"
+
+func NewClient(baseUrl, apiKey string) *client {
+	return &client{
+		baseUrl: baseUrl + "/api/%s&apikey=" + apiKey,
+	}
+}
+
+func NewEthereumClient(apiKey string) *client {
+	return NewClient("https://api.etherscan.io", apiKey)
+}
+
+func NewOptimismClient(apiKey string) *client {
+	return NewClient("https://api-optimistic.etherscan.io", apiKey)
+}
+
+func (c *client) fetch(url string) ([]byte, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return body, nil
+}
+
+func (c *client) fetchEtherscanApi(url string) (apiResponse, error) {
+	return retry.Do[apiResponse](context.Background(), apiMaxRetries, retry.Fixed(apiRetryDelay), func() (apiResponse, error) {
+		body, err := c.fetch(url)
+		if err != nil {
+			return apiResponse{}, err
+		}
+
+		var response apiResponse
+		err = json.Unmarshal(body, &response)
+		if err != nil {
+			return apiResponse{}, fmt.Errorf("failed to unmarshal as apiResponse: %w", err)
+		}
+
+		if response.Message != "OK" {
+			var resultString string
+			err = json.Unmarshal(response.Result, &resultString)
+			if err != nil {
+				return apiResponse{}, fmt.Errorf("failed to unmarshal response.Result as a string: %w", err)
+			}
+
+			if resultString == errRateLimited {
+				return apiResponse{}, errors.New(strings.ToLower(errRateLimited))
```

Comment:
I converted to lowercase because the linter complains that the error starts with an uppercase letter
<img width="810" alt="image" src="https://github.com/ethereum-optimism/optimism/assets/20375610/5136c739-ec85-4553-8f0e-c3ef3c46e20f">